### PR TITLE
Review

### PR DESCRIPTION
**Pull request by tdpauw**
Hey 👋,

Here are some proposed changes. This is mainly based on how I like to organise projects and how I have seen projects organised as well as on Linux standards and conventions.

I've tried to explain the rationale in the commit bodies.

What I try to achieve is discoverability when browsing a directory structure and when executing `make`.

You might like to have a look at the [Linux Standard Directory Structure](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) to understand why I've moved some files to other locations inside the Docker image.

There is still room for improvement:
- Using the slim version of Debian or even using Alpine Linux. This is a hack to reduce the Docker image size even further

I hope this helps.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,17 +1,27 @@
-all: 
-	@docker-compose -f ./srcs/docker-compose.yml up
+.DEFAULT_GOAL := help
```
Comment (mlrcbsousa) [inline Makefile:1]: So I tried reading up and understand this `DEFAULT_GOAL` is it that when I run `make` without any argument or a wrong argument it will default to `help`?

File: Makefile [hunk 2]
```diff
@@ -1,17 +1,27 @@
...
+	@docker-compose build
+
+clean: ## Deletes all containers, images, volumes and networks
+	@docker stop $$(docker ps -qa --filter 'name=inception-*');\
+	docker rm $$(docker ps -qa --filter 'name=inception-*');\
+	docker rmi -f $$(docker images "inception_*" -qa);\
+	docker volume rm $$(docker volume ls -q --filter 'name=inception-*');\
+	docker network rm $$(docker network ls -q --filter 'name=inception-*');\
+
+init: ## Initialisation
```
Comment (mlrcbsousa) [inline Makefile:21]: yes perfect, makes sense, init -> build -> up

File: Makefile [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
+down: ## Stops the inception app
+	@docker-compose down
 
-.PHONY: all re down clean
\ No newline at end of file
+build: ## Builds the inception docker images
+	@docker-compose build
+
+clean: ## Deletes all containers, images, volumes and networks
+	@docker stop $$(docker ps -qa --filter 'name=inception-*');\
```
Comment (mlrcbsousa) [inline Makefile:15]: ah ok so more specific filters in order to not delete everything on the system

File: Makefile [hunk 4]
```diff
@@ -1,17 +1,27 @@
...
+	docker volume rm $$(docker volume ls -q --filter 'name=inception-*');\
+	docker network rm $$(docker network ls -q --filter 'name=inception-*');\
+
+init: ## Initialisation
+	mkdir -p _data/mariadb _data/wordpress
+
+.PHONY: all build up down clean init
+
+help:
+	@grep -h -E '^[a-zA-Z0-9_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
Comment (mlrcbsousa) [inline Makefile:27]: ah ok this is great, haha so cool

File: docker-compose.yml [hunk 5]
```diff
@@ -2,29 +2,25 @@ version: "3.5"
...
 
 volumes:
   wp:
     driver: local
-    name: wp
     driver_opts:
       type: none
       o: bind
-      device: ~/inception/wordpress
+      device: ./_data/wordpress
```
Comment (mlrcbsousa) [inline docker-compose.yml:13]: yes this is perfect, I was struggling to know where to make these folders and the problem of running the compose command and not having the folders yet, really cool thanks

File: docker-compose.yml [hunk 6]
```diff
@@ -2,29 +2,25 @@ version: "3.5"
 
 networks:
   app-network:
-    name: app-network
```
Comment (mlrcbsousa) [inline docker-compose.yml:5]: why did you remove this?

File: docker-compose.yml [hunk 7]
```diff
@@ -2,29 +2,25 @@ version: "3.5"
...
-    name: db
     driver_opts:
       type: none
       o: bind
-      device: ~/inception/mariadb
+      device: ./_data/mariadb
 
 services:
   mariadb:
-    container_name: mariadb
```
Comment (mlrcbsousa) [inline docker-compose.yml:26]: ok so `container_name` has no value added

File: src/wordpress/bin/entrypoint.sh [hunk 8]
```diff
@@ -14,7 +14,7 @@ if [ ! -f /var/www/html/wp-config.php ]; then
 	mv wp-cli.phar /usr/local/bin/wp;
 	cd /var/www/html;
 	wp core download --allow-root;
-	mv /var/www/wp-config.php /var/www/html/
+	mv /var/lib/wp/wp-config.php /var/www/html/
```
Comment (mlrcbsousa) [inline src/wordpress/bin/entrypoint.sh:17]: so here using this directory makes it clearer the intent of this file, is that it?

File: src/wordpress/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,17 @@
+FROM debian:buster
+
+RUN apt-get update && \
+  apt-get --yes --no-install-recommends install php7.3 php-mysqli php-fpm wget sendmail \
+  && rm -rf /var/lib/apt/lists/*
```
Comment (mlrcbsousa) [inline src/wordpress/Dockerfile:5]: What does the remove do?

Also is it convention to use one RUN command?

File: src/wordpress/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+RUN mkdir /var/lib/wp
+
+EXPOSE 9000
+
+COPY ./conf/www.conf /etc/php/7.3/fpm/pool.d
+COPY ./bin /usr/local/bin
```
Comment (mlrcbsousa) [inline src/wordpress/Dockerfile:12]: I'm confused with this line 😅

File: Makefile (hunk 1 above)
Comment (tdpauw) [reply Makefile:1]: `DEFAULT_GOAL` sets the default goal/target when no is given via the commandline, i.e. as you say when only executing:

```
make
```

if no `DEFAULT_GOAL` is defined it takes the first goal/target from the Makefile.

Here it will default to `help`.

File: Makefile (hunk 3 above)
Comment (tdpauw) [reply Makefile:15]: yes indeed, this avoids you from deleting existing containers or volumes you might have on your system (as it is the case with me).

File: docker-compose.yml (hunk 6 above)
Comment (tdpauw) [reply docker-compose.yml:5]: because docker-compose will name everything `projectname-resource`, where resource is in this case `app-network`.

```
networks:
  app-network:
    ...
```

So you will have: `inception-app-network`.

The prefixing allows to filter what to delete in the `clean` Makefile target.

File: src/wordpress/bin/entrypoint.sh (hunk 8 above)
Comment (tdpauw) [reply src/wordpress/bin/entrypoint.sh:17]: yes, and it is also a Linux convention to place application-specific data in `/var/lib`.

I've left `/var/www/html` not knowing how WordPress works. I assume WordPress picks up everything located in `/var/www/html` and executes it.

File: src/wordpress/Dockerfile (hunk 9 above)
Comment (tdpauw) [reply src/wordpress/Dockerfile:5]: the removal cleans the `apt-get` cache

the idea of combining commands in one `RUN` is to limit the Docker image layers.
Each Dockerfile command creates a layer which is essentially a Docker image on its own. Each layer builds on top of the previous layer. So each layer stores the changes to the image it is based on.

This has specifically some side effects when removing files. The files are removed from the current layer but not from the previous layer.  As a result, this adds up to the total Docker image size.

The layers are used for Docker image build caching. As long as a Dockerfile line has not changed, the corresponding layer is not rebuilt. If a layer needs rebuilding, all following layers are rebuilt. This has a consequence on how to write a Dockerfile.
Usually, the things that change infrequently are put at the top (like installing Ubuntu packages, OS set up, ...) and the things that change frequently (source code) are placed at the bottom.
=> Dockerfile organisation:
1. install packages
2. configuration
3. add source code or binaries for compiled languages

[This](https://dzone.com/articles/docker-layers-explained) article explains the concept of Docker image layers more in detail.

File: src/wordpress/Dockerfile (hunk 10 above)
Comment (tdpauw) [reply src/wordpress/Dockerfile:12]: by which line are you confused?

Comment (mlrcbsousa) [reply src/wordpress/Dockerfile:12]: `COPY ./bin /usr/local/bin`

I guess here you are making sure the binaries are the same in the container but don't we want to use the ones installed in the container

File: src/wordpress/Dockerfile (hunk 9 above)
Comment (mlrcbsousa) [reply src/wordpress/Dockerfile:5]: This is very cool and interesting thank you!

File: src/wordpress/Dockerfile (hunk 10 above)
Comment (tdpauw) [reply src/wordpress/Dockerfile:12]: Oh, this copies the content of the `./src/wordpress/bin` folder to the `/usr/local/bin` of the Docker image.
`/usr/local/bin` may not have a trailing `/`, otherwise it copies everything at `/usr/local/bin/bin`.
However, I do not think it overwrites the content of `/usr/local/bin`. Anyhow, most of the time `/usr/local/bin` is empty. Linux distributions should not install anything there. It is supposed to be used for custom installations like here.
